### ISSUE TITLE
Accessibility support for embedded Android views

### ISSUE THREAD

**Issue by amirh** [flutter/flutter#19418]

**Comment by jonahwilliams** [comment]
This requires a new SemanticsFlag, probably something like `isExternalView`.  The semantics node which contained this flag could only have this flag and a label, which would be the serialized identifier used by the native view registry.

**Comment by jonahwilliams** [comment]
Per offline discussion: It is not currently possible to determine the correct parent-child relationship for an Android view with synthetic (not based on a View hierarchy) accessibility tree, such as a web view. The nodes can be obtained, there just isn't anything we can do with them.

There may be another API we need to provide to embedded views to communicate 1) what the full accessibility node hierarchy is and 2) when to update it/receive events.

Reassigning to @amirh for planning.

**Comment by mehmetf** [comment]
@amirh Could you please update this bug? This is one of the primary issues for customer:dream on webview.

**Comment by amirh** [comment]
I'm investigating a few potential approaches, nothing concrete to share yet.

**Comment by amirh** [comment]
An e2e working prototype that works up to Android O is available here: https://github.com/amirh/engine/tree/a11y_hacks

I'm starting to work on landing this solution in the engine. We are still exploring solutions for Android P and later.

**Comment by amirh** [comment]
With the list of referenced PRs that landed platform accessibility is working for platform views with a pure virtual accessibility nodes tree(e.g WebView) on Android versions up to O.

There's still work to get this issue fully resolved, but we do have what "customer: dream" needs so I'm removing the label.

**Comment by mklim** [comment]
Landed a workaround, flutter/engine#9321, that should enable a11y in embedded Android views that use virtual node hierarchies for all Android versions.

I've been doing some deep dives into enabling a11y in all embedded Android views, not just the ones that use virtual node hierarchies. Ideally we'd like to do this by reparenting the embedded view's a11y node tree under Flutter's a11y node tree, but so far I haven't figured out a way to make this work.

Basically, in Android a11y nodes can be "real" (created by Android `View` classes themselves) or "virtual" (generated by application code via creating an `AccessibilityNodeProvider`). Flutter nodes are virtual, created with its own A11yNodeProvider since none of its UI elements are "real" Android views and the whole thing rolls down to a texture. When we embed a `View` into our Flutter UI using an `AndroidView`, we achieve this on the Android side by creating a `VirtualDisplay` and throwing the `View` into that. This makes sense. There’s no real parent for it to directly belong to, since our widget tree exists entirely inside of Flutter’s framework logic and has no representation on the Android platform other than the final rasterized output.

But this lack of `View` hierarchy causes breakdowns in trying to associate a "real" View's nodes with our tree. a11y code in Android frequently relies on the View hierarchy for state management and querying extra information, so attempting to parent the child’s hierarchy from the embedded view to our virtual hierarchy fails to really correctly associate the underlying view hierarchy in a functional way for a11y. In other words, reparenting the a11y node trees _only_ fixes the a11y node tree themselves, but there are still multiple places where Android a11y code relies on being able to query the “real” view hierarchy for required info and ends up failing with bugs in our case.

**Comment by BugsBunnyBR** [comment]
Hi, 

1) Is there a plan to remove the reflection calls to Android framework private API calls? 

2) Any strategy to enable accessibility for new Android versions?

If there is no current clear path:
3) Wouldn't it be a reasonable idea to file a request feature for android framework team to expose those methods in a the `AccessibilityNodeInfoCompat` or something similar ? 

4) Would any more changes be needed to address accessibility for Flutter? Like the ability to reparent the android view's a11y in Flutter node tree @mklim brings in this thread.  Would it be possible to be filled to the android framework team?

I am not particularly familiar with the implementation problems or with accessibility, but I do know the reflection based solution will not work in new Android versions, so if someone that can describe the needs and issues could file a feature request here https://issuetracker.google.com/issues/new?component=460834 it might be helpful in archiving a good accessibility support to Flutter.

Sorry if this is already discussed offline or in other issues here, I just felt this comment could help.
Thats all folks. :-)

**Comment by mklim** [comment]
> Any strategy to enable accessibility for new Android versions?
> I do know the reflection based solution will not work in new Android versions

The reflection we're using is currently [light greylisted](https://developer.android.com/distribute/best-practices/develop/restrictions-non-sdk-interfaces#list-names), so it works even on newer API versions. You're totally right that this could change in the future with additional methods getting blacklisted over time, and is just generally brittle. But this should already be working across new Android versions now. If you see this behaving badly on a new version of Android, please file a bug for it or call it out here since that's unexpected.

In an absolute worst case scenario, like a future version of Android blacklists the APIs that we're using, we still would have one last option. It's possible to get the data that we need by reading an a11y node's parceled data. That's worse than relying on reflection in some ways so I wouldn't move to it unless we absolutely had to, but at least there is that fallback.

For 1, 3, and 4: 

We're in discussions to have a better API for us in the future, stay tuned!

> Sorry if this is already discussed offline or in other issues here, I just felt this comment could help.

No problem, thanks for the input. :)

**Comment by BugsBunnyBR** [comment]
@mklim , thanks for the reply.
Maybe this https://github.com/flutter/flutter/issues/30395 is outdated or I don't get the big picture, but  this issue referes to PRs disabling the reflective calls for Android O and new android versions.

> 
We're in discussions to have a better API for us in the future, stay tuned!

I am trying :-D

**Comment by mklim** [comment]
@BugsBunnyBR  no problem! Ah yeah, it looks like the full story here is spread out across a couple different places. So originally this was addressed using reflective calls that were completely blacklisted post O, causing #30395 and meaning a11y functionality didn't work at all post O and that post O there were also errors about the reflection being used, #30395. That was closed by disabling the calls that were blacklisted when running post O, but still with no a11y functionality on those versions. Then after that flutter/engine#9321 was merged in, which put in some workarounds that add a11y functionality in post O for the first time but are not ideal long term solutions. That's still the state of the issue now. I just went and filed [an issue on the Android SDK tracker](https://issuetracker.google.com/issues/138442751) for the underlying Android issue too.

**Comment by blasten** [comment]
This is automatically fixed by hybrid composition since the view is placed in the Android view hierarchy.

**Comment by chinmaygarde** [comment]
@blasten Can you confirm that this issue has been addressed already or update the milestone?

**Comment by chinmaygarde** [comment]
Per Ray, this should be fixed in 1.22. That is under verification right now.

**Comment by blasten** [comment]
This is fixed in 1.22. @bparrishMines PR should be the last step prior to closing this issue.

**Comment by chinmaygarde** [comment]
Can you update the milestone of this issue please?

**Comment by blasten** [comment]
@chinmaygarde added label `waiting for PR to land (fixed)`

**Comment by blasten** [comment]
This has been fixed in 1.22 - use hybrid composition to make an Android view fully accessible.  Hybrid composition should work 100% of the times just like any other Android view inside a vanilla Android app.

https://flutter.dev/docs/development/platform-integration/platform-views?tab=android-platform-views-kotlin-tab#hybrid-composition
